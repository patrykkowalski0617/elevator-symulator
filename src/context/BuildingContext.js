import React, { createContext, useState } from "react";

export const BuildingContext = createContext();

const BuildingContextProvider = props => {
    const [numberOfFloors, setNumberOfFloors] = useState(8);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
