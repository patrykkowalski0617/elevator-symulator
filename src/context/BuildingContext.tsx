import React, { createContext, useState } from "react";

interface IContextProps {
    numberOfFloors: number;
    setNumberOfFloors: (n: number) => void;
    numberOfCars: number;
    setNnmberOfCars: (n: number) => void;
    automotion: boolean;
    setAutomotion: (b: boolean) => void;
    creatingStickMan: boolean;
    setCreatingStickMan: (arr: boolean) => void;
}

export const BuildingContext = createContext({} as IContextProps);

const BuildingContextProvider = (props: { children: React.ReactNode }) => {
    const [numberOfFloors, setNumberOfFloors] = useState(11);
    const howManyCars = Math.floor(numberOfFloors / 5);
    const [numberOfCars, setNnmberOfCars] = useState(howManyCars);
    const [automotion, setAutomotion] = useState(false);
    const [creatingStickMan, setCreatingStickMan] = useState<boolean>(false);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors,
                numberOfCars,
                setNnmberOfCars,
                automotion,
                setAutomotion,
                creatingStickMan,
                setCreatingStickMan
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
