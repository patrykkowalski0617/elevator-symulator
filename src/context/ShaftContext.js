import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const ShaftContext = createContext();

const ShaftContextProvider = props => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = val => Array(numberOfCars).fill(val);

    const [carCurrenFloorCon, setCarCurrenFloorCon] = useState(initArr(0)); // next floor on way or floor that has been just reached, but not floor which has been just left

    // console.log(carCurrenFloorCon);

    return (
        <ShaftContext.Provider
            value={{
                carCurrenFloorCon,
                setCarCurrenFloorCon
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
