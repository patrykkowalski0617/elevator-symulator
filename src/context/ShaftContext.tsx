import React, { createContext, useState, useContext, useEffect } from "react";
import { BuildingContext } from "./BuildingContext";

interface IContextProps {
    allCarsCurrentFloor: number[];
    updateCarCurrentFloor: (carId: number, currentFloor: number) => void;
    allCarsFloorAssignments: number[][];
    addCarFloorAssignment: (carId: number, floorNumber: number) => void;
    removeCarFloorAssignment: (carId: number, floorNumber: number) => void;
    allCarsState: string[];
    updateCarState: (carId: number, state: string) => void;
    allCarsStickMans: {
        lifeState: string;
        destination: number;
        carId: number | null;
        placeInCar: number | null;
    }[][];
    addPassengers: (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => void;
    allCarsDirection: string | null[];
    updateCarDirection: (carId: number, direction: string) => void;
}

export const ShaftContext = createContext({} as IContextProps);

const ShaftContextProvider = (props: { children: React.ReactNode }) => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = (val: any) => Array(numberOfCars).fill(val);

    // next floor on way or floor that has been just reached, but not floor which has been just left
    const [allCarsCurrentFloor, setAllCarsCurrentFloor] = useState(initArr(0));
    // ready, go-up, go-down, go-up-door-open, go-down-door-open
    const [allCarsState, setAllCarsState] = useState(initArr("ready"));
    const [allCarsFloorAssignments, setAllCarsFloorAssignments] = useState(
        initArr([])
    );
    const [allCarsStickMans, setAllCarsStickMans] = useState(initArr([]));
    const [allCarsDirection, setAllCarsDirection] = useState(initArr(null));

    const updateCarCurrentFloor = (carId: number, currentFloor: number) => {
        const _allCarsCurrentFloor = [...allCarsCurrentFloor];
        _allCarsCurrentFloor.splice(carId, 1, currentFloor);
        setAllCarsCurrentFloor([..._allCarsCurrentFloor]);
    };

    const updateCarState = (carId: number, state: string) => {
        const _allCarsState = [...allCarsState];
        _allCarsState.splice(carId, 1, state);
        setAllCarsState([..._allCarsState]);
    };

    const addCarFloorAssignment = (carId: number, floorNumber: number) => {
        if (carId !== null) {
            const _allCarsFloorAssignments = [...allCarsFloorAssignments];
            const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
            const carFloorAssignmentUpdated = [
                ...carFloorAssignment,
                floorNumber
            ];
            _allCarsFloorAssignments.splice(
                carId,
                1,
                carFloorAssignmentUpdated
            );
            setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
        }
    };

    const [_allCarsStickMans, set_allCarsStickMans] = useState(
        allCarsStickMans
    );
    useEffect(() => {
        set_allCarsStickMans(allCarsStickMans);
    }, [allCarsStickMans]);
    const addPassengers = (
        data: {
            destination: number;
            direction: string | null;
            lifeState: string;
            carId: number | null;
            placeInCar: number | null;
        }[]
    ) => {
        if (data[0]) {
            const carId =
                typeof data[0].carId === "number" ? data[0].carId : -1;
            if (allCarsStickMans[carId].length < 4) {
                const carPassengers = [..._allCarsStickMans[carId], ...data];
                _allCarsStickMans.splice(carId, 1, carPassengers);
                setAllCarsStickMans([..._allCarsStickMans]);
            } else {
                console.warn("car is full!");
            }
        } else {
            console.warn(
                "there are no stickmen on this floor who have the same direction as the car!"
            );
        }
    };

    const removeCarFloorAssignment = (carId: number, floorNumber: number) => {
        const _allCarsFloorAssignments = [...allCarsFloorAssignments];
        const carFloorAssignment = [..._allCarsFloorAssignments[carId]];
        const index = carFloorAssignment.indexOf(floorNumber);
        carFloorAssignment.splice(index, 1);
        const carFloorAssignmentUpdated = [...carFloorAssignment];
        _allCarsFloorAssignments.splice(carId, 1, carFloorAssignmentUpdated);
        setAllCarsFloorAssignments([..._allCarsFloorAssignments]);
    };

    const updateCarDirection = (carId: number, direction: string) => {
        const _allCarsDirection = allCarsDirection;
        _allCarsDirection.splice(carId, 1, direction);
        setAllCarsDirection([..._allCarsDirection]);
    };

    return (
        <ShaftContext.Provider
            value={{
                allCarsCurrentFloor,
                updateCarCurrentFloor,
                allCarsFloorAssignments,
                addCarFloorAssignment,
                removeCarFloorAssignment,
                allCarsState,
                updateCarState,
                allCarsStickMans,
                addPassengers,
                allCarsDirection,
                updateCarDirection
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
