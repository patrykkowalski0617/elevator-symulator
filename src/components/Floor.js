import React, { useState, useContext } from "react";
import styled from "styled-components";
import { ShaftContext } from "../context/ShaftContext";

const FloorStyled = styled.div`
    border-bottom: 4px solid;
    height: ${props => props.floorHeight}px;
`;

const CarLight = styled.div`
    background: orange;
    width: 7px;
    height: 7px;
    margin: 7px 7px auto auto;
    float: right;
    border-radius: 100%;
`;

export default function Floor({ floorHeight, floorNumber, role }) {
    const [waitingForCar, setWaitingForCar] = useState(false);
    const { carCurrenFloorCon } = useContext(ShaftContext);

    const theNearestCarNum = arrData => {
        let min = Number.POSITIVE_INFINITY;
        let theNearestCarNum = null;

        for (let i = 0; i < arrData.length; i++) {
            let distanceToFloor = arrData[i].distanceToFloor;
            let carNumber = arrData[i].carNumber;

            if (distanceToFloor < min) {
                theNearestCarNum = carNumber;
            }
            min = distanceToFloor < min ? distanceToFloor : min;
        }

        return theNearestCarNum;
    };

    const onClickHandler = e => {
        if (!waitingForCar) {
            setWaitingForCar(true);
            const distancesToCars = carCurrenFloorCon.map((item, index) => {
                return {
                    carNumber: index,
                    distanceToFloor: Math.abs(floorNumber - item)
                };
            });

            // DONE:
            // find the nearest Car
            console.log(theNearestCarNum(distancesToCars));

            // TO OD:
            //  get elevators with correct direction
            // get elevator whitch is not busy or is less bussy than other
        }
    };

    return (
        <FloorStyled floorHeight={floorHeight}>
            {role === "enter-floor" ? (
                <button
                    data-floor-number={floorNumber}
                    onClick={e => {
                        onClickHandler(e);
                    }}
                >
                    Call car
                </button>
            ) : (
                ""
            )}
            {waitingForCar ? <CarLight></CarLight> : ""}
        </FloorStyled>
    );
}
