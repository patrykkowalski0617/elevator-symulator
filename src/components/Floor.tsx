import React, { useState, useContext, useEffect } from "react";
import styled from "styled-components";
import { ShaftContext } from "../context/ShaftContext";
import { FloorsContext } from "../context/FloorsContext";
import { theNearestCar } from "../car_logic";

const FloorStyled = styled.div<{ numberOfFloors: number; floorColor: string }>`
    height: ${props => 100 / props.numberOfFloors}%;
    border-style: solid;
    border-color: #222;
    border-width: 2px 0;
    background-color: ${props => props.floorColor};
`;

const CarLight = styled.div`
    background: orange;
    width: 7px;
    height: 7px;
    margin: 7px 7px auto auto;
    float: right;
    border-radius: 100%;
`;

const AssignedCar = styled.div`
    float: right;
    margin: 2px 7px;
`;

type FloorProps = {
    floorNumber: number;
    numberOfFloors: number;
    role: string;
    floorColor: string;
};

const Floor = ({
    floorNumber,
    numberOfFloors,
    role,
    floorColor
}: FloorProps) => {
    const {
        allCarsState,
        allCarsCurrentFloor,
        addCarFloorAssignment,
        allCarsFloorAssignments
    } = useContext(ShaftContext);
    const { floorsWaitingForCar, addFloorWaitingForCar } = useContext(
        FloorsContext
    );
    const waitingForCar = floorsWaitingForCar[floorNumber];

    const [noCar, setNoCar] = useState(false);
    const [assignedCar, setAssignedCar] = useState<number | null>(null);

    const call = () => {
        addFloorWaitingForCar(floorNumber);
    };

    useEffect(() => {
        if (waitingForCar) {
            const carId = theNearestCar({
                allCarsState,
                allCarsCurrentFloor,
                allCarsFloorAssignments,
                floorNumber
            });
            if (carId !== null && carId !== undefined && carId >= 0) {
                addCarFloorAssignment(carId, floorNumber);
                setAssignedCar(carId);
            } else {
                setNoCar(true);
            }
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [waitingForCar]);

    return (
        <FloorStyled numberOfFloors={numberOfFloors} floorColor={floorColor}>
            {role === "enter-floor" ? (
                <>
                    <button
                        style={noCar ? { background: "red" } : {}}
                        data-floor-number={floorNumber}
                        onClick={call}
                    >
                        {!noCar ? "Call car" : "No available cars"}
                    </button>
                    {waitingForCar ? (
                        <>
                            <CarLight></CarLight>
                            <AssignedCar>
                                {assignedCar !== null
                                    ? `Car: ${assignedCar}`
                                    : ""}
                            </AssignedCar>
                        </>
                    ) : null}
                </>
            ) : null}

            {role === "exit-floor" ? <p>{floorNumber}</p> : null}
        </FloorStyled>
    );
};

export default Floor;
