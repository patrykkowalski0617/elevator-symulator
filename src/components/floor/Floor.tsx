import React, { useState, useContext, useEffect } from "react";
import {
    FloorStyled,
    AssignedCars,
    CarInfo,
    CarInfoItem,
    CreateBtn,
    Light
} from "./FloorStyled";
import { ShaftContext } from "../../context/ShaftContext";
import { FloorsContext } from "../../context/FloorsContext";
import { BuildingContext } from "../../context/BuildingContext";
import StickMan from "../stickman/StickMan";
import { waitingForCarUpdate, callCar } from "./logic";

type FloorProps = {
    floorNumber: number;
    numberOfFloors: number;
    floorColor: string;
};

const Floor = ({ floorNumber, numberOfFloors, floorColor }: FloorProps) => {
    const {
        allCarsState,
        allCarsCurrentFloor,
        addCarFloorAssignment,
        allCarsFloorAssignments,
        allCarsStickMansDestinations,
        allCarsDirection,
        addPassengers
    } = useContext(ShaftContext);
    const {
        floorsWaitingForCar,
        addFloorWaitingForCar,
        allFloorsStickMansDestinations
    } = useContext(FloorsContext);
    const stickMansDestinations: number[] =
        allFloorsStickMansDestinations[floorNumber];
    const waitingForCar: { up: boolean; down: boolean } =
        floorsWaitingForCar[floorNumber];
    const { setCreatingStickMan, numberOfCars } = useContext(BuildingContext);

    const [noCar, setNoCar] = useState<boolean>(false);
    const [assignedCars, setAssignedCars] = useState<number[]>([]);
    const [carsOnFloor, setCarsOnFloor] = useState<number[]>([]);
    const [indexesForGetIn, setIndexesForGetIn] = useState<number[][]>(
        Array(numberOfCars).fill([])
    );

    useEffect(() => {
        let assignedCar: number;
        let stickmans: number[];
        let direction: string | null;
        let freePlacesInCar: number;
        let firstInQueue: number;

        for (let i = 0; i < allCarsState.length; i++) {
            const state = allCarsState[i];
            const currentFloor = allCarsCurrentFloor[i];
            if (
                floorNumber === currentFloor &&
                state.includes("door-open") &&
                !carsOnFloor.includes(i)
            ) {
                assignedCar = i;
                // console.log("car " + i + " came");
                setTimeout(() => {
                    freePlacesInCar =
                        4 - allCarsStickMansDestinations[assignedCar].length;
                    stickmans = stickMansDestinations;
                    firstInQueue =
                        stickMansDestinations[stickMansDestinations.length - 1];
                    direction = allCarsDirection[assignedCar]
                        ? allCarsDirection[assignedCar]
                        : firstInQueue > floorNumber
                        ? "up"
                        : firstInQueue < floorNumber
                        ? "down"
                        : null;
                    const indexesGetIn: number[] = [];
                    for (let j = 0; j < stickmans.length; j++) {
                        const stickman = stickmans[j];
                        if (indexesGetIn.length < freePlacesInCar) {
                            if (direction === "up" && stickman > floorNumber) {
                                indexesGetIn.push(j);
                            } else if (
                                direction === "down" &&
                                stickman < floorNumber
                            ) {
                                indexesGetIn.push(j);
                            }
                        }
                    }
                    const _indexesForGetIn = indexesForGetIn;
                    _indexesForGetIn.splice(assignedCar, 1, indexesGetIn);
                    setIndexesForGetIn([..._indexesForGetIn]);
                    // console.log("get in", {
                    //     freePlacesInCar,
                    //     stickmans,
                    //     firstInQueue,
                    //     indexesForGetIn,
                    //     direction,
                    //     assignedCar
                    // });
                    setTimeout(() => {
                        // addPassengers(i, 0);
                        // console.log("replace stickmans");
                    }, 1000);
                    setCarsOnFloor([...carsOnFloor, i]);
                }, 1000);
            }
        }
    }, [allCarsState]);

    useEffect(() => {
        waitingForCarUpdate({
            waitingForCar,
            stickMansDestinations,
            addFloorWaitingForCar,
            floorNumber
        });
    }, [stickMansDestinations]);

    const _callCar = callCar({
        allCarsState,
        allCarsCurrentFloor,
        allCarsFloorAssignments,
        floorNumber,
        addCarFloorAssignment,
        setAssignedCars,
        assignedCars,
        setNoCar
    });

    useEffect(() => {
        if (waitingForCar.up) {
            _callCar();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [waitingForCar.up]);

    useEffect(() => {
        if (waitingForCar.down) {
            _callCar();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [waitingForCar.down]);

    const createStickmanHandler = () => {
        setCreatingStickMan(floorNumber);
    };

    const stickMans = stickMansDestinations.map(
        (item: number, index: number) => {
            let getIn: boolean = false;
            let assignedCar: number | null = null;
            for (let i = 0; i < indexesForGetIn.length; i++) {
                if (indexesForGetIn[i].includes(index)) {
                    getIn = true;
                    assignedCar = i;
                }
            }

            return (
                <StickMan
                    key={index}
                    stickId={index}
                    destination={item}
                    getIn={getIn}
                    numberOfPassengers={indexesForGetIn.length}
                    assignedCar={assignedCar}
                    place={"floor"}
                />
            );
        }
    );

    return (
        <FloorStyled numberOfFloors={numberOfFloors} floorColor={floorColor}>
            <CarInfo>
                {waitingForCar.up ||
                (waitingForCar.down && assignedCars.length) ? (
                    <AssignedCars>Cars: {assignedCars.join(", ")}</AssignedCars>
                ) : null}
                <CarInfoItem>
                    <Light waitingForCar={waitingForCar} noCar={noCar}></Light>
                </CarInfoItem>
                <CarInfoItem>
                    <CreateBtn onClick={createStickmanHandler}>+</CreateBtn>
                </CarInfoItem>
            </CarInfo>
            {stickMans}
        </FloorStyled>
    );
};

export default Floor;
