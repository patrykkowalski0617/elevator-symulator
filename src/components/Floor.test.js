import React from "react";
import Floor from "./Floor";
import renderer from "react-test-renderer";
import ShaftContext from "../context/ShaftContext";
import BuildingContext from "../context/BuildingContext";

test("", () => {
    const tree = renderer.create(
        <BuildingContext>
            <ShaftContext>
                <Floor title="tiiitle"></Floor>
            </ShaftContext>
        </BuildingContext>
    );
    expect(tree.toJSON()).toMatchSnapshot();
});
