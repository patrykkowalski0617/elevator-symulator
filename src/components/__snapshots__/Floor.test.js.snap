// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="sc-AxjAm kofykE"
  title="tiiitle"
/>
`;
